File: src/windows/AmbientContextMcp.Core/Hub/LocalContextCursorTracker.cs
using System.Text;

namespace AmbientContextMcp.Core.Hub;

public sealed class LocalContextCursorTracker
{
    // ClientId ごとの位置を永久蓄積するとリークになるので、一定期間アクセスのない
    // entry は古い順に落とす。MCP クライアントは通常 1 ホスト数個だが、UUID で
    // 毎回名乗る misbehaving client がいた場合に dict が無制限に膨らむのを防ぐ。
    public static readonly TimeSpan DefaultStaleClientTtl = TimeSpan.FromDays(7);

    private readonly Dictionary<string, Entry> _clientPositions = new(StringComparer.OrdinalIgnoreCase);

    public CursorResult Resolve(
        string clientId,
        string cursor,
        bool isHistoryQuery,
        long firstSequence,
        long latestSequence,
        DateTimeOffset now)
    {
        if (TryDecode(cursor, out var cursorSequence))
        {
            return new CursorResult(
                ClampExpired(cursorSequence, firstSequence),
                IsExpired(cursorSequence, firstSequence));
        }

        if (isHistoryQuery)
        {
            return new CursorResult(firstSequence == 0 ? 0 : Math.Max(0, firstSequence - 1), false);
        }

        var normalizedClientId = NormalizeClientId(clientId);
        if (_clientPositions.TryGetValue(normalizedClientId, out var entry))
        {
            // touch: 読み出しでも lastSeen を更新しないと、subscribe しっぱなしで
            // Advance だけ呼ばれないパス (= 取得 0 件で位置が進まない) が TTL で落ちる。
            _clientPositions[normalizedClientId] = entry with { LastSeen = now };
            return new CursorResult(
                ClampExpired(entry.Sequence, firstSequence),
                IsExpired(entry.Sequence, firstSequence));
        }

        _clientPositions[normalizedClientId] = new Entry(latestSequence, now);
        return new CursorResult(latestSequence, false);
    }

    public void Advance(string clientId, long sequence, DateTimeOffset now)
    {
        _clientPositions[NormalizeClientId(clientId)] = new Entry(sequence, now);
    }

    /// <summary>
    /// 最終アクセスから ttl 経過した client entry を削除する。戻り値は削除件数。
    /// </summary>
    public int PruneStale(DateTimeOffset now, TimeSpan? ttl = null)
    {
        var threshold = now - (ttl ?? DefaultStaleClientTtl);
        List<string>? stale = null;
        foreach (var pair in _clientPositions)
        {
            if (pair.Value.LastSeen < threshold)
            {
                (stale ??= []).Add(pair.Key);
            }
        }

        if (stale is null)
        {
            return 0;
        }

        foreach (var key in stale)
        {
            _clientPositions.Remove(key);
        }

        return stale.Count;
    }

    public int TrackedClientCount => _clientPositions.Count;

    public static string Encode(long sequence)
    {
        return Convert.ToBase64String(Encoding.UTF8.GetBytes("seq:" + sequence))
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }

    private static long ClampExpired(long sequence, long firstSequence)
    {
        return IsExpired(sequence, firstSequence)
            ? Math.Max(0, firstSequence - 1)
            : sequence;
    }

    private static bool IsExpired(long sequence, long firstSequence)
    {
        return firstSequence > 0 && sequence < firstSequence - 1;
    }

    private static string NormalizeClientId(string clientId)
    {
        return string.IsNullOrWhiteSpace(clientId) ? "anonymous" : clientId.Trim();
    }

    private static bool TryDecode(string cursor, out long sequence)
    {
        sequence = 0;
        if (string.IsNullOrWhiteSpace(cursor))
        {
            return false;
        }

        try
        {
            var padded = cursor.Replace('-', '+').Replace('_', '/');
            padded = padded.PadRight(padded.Length + ((4 - padded.Length % 4) % 4), '=');
            var text = Encoding.UTF8.GetString(Convert.FromBase64String(padded));
            return text.StartsWith("seq:", StringComparison.OrdinalIgnoreCase) &&
                   long.TryParse(text[4..], out sequence);
        }
        catch
        {
            return false;
        }
    }

    public readonly record struct CursorResult(long Sequence, bool Expired);

    private readonly record struct Entry(long Sequence, DateTimeOffset LastSeen);
}

File: src/macos/Sources/AmbientContextCore/Hub/LocalContextCursorTracker.swift
import Foundation

/// ClientId ごとの cursor 位置を保持する。
/// 位置を永久蓄積するとリークになるので、一定期間アクセスのない entry は落とす。
/// MCP クライアントは通常 1 ホスト数個だが、UUID で毎回名乗る misbehaving client がいた場合に
/// dict が無制限に膨らむのを防ぐ。
public final class LocalContextCursorTracker {
    public static let defaultStaleClientTtl: TimeInterval = 7 * 24 * 60 * 60

    public struct CursorResult: Sendable, Hashable {
        public var sequence: Int64
        public var expired: Bool

        public init(sequence: Int64, expired: Bool) {
            self.sequence = sequence
            self.expired = expired
        }
    }

    private struct Entry {
        var sequence: Int64
        var lastSeen: Date
    }

    /// キーは正規化 (大文字小文字無視) 済み clientId。
    private var clientPositions: [String: Entry] = [:]

    public init() {}

    public func resolve(
        clientId: String,
        cursor: String,
        isHistoryQuery: Bool,
        firstSequence: Int64,
        latestSequence: Int64,
        now: Date
    ) -> CursorResult {
        if let cursorSequence = Self.decode(cursor) {
            return CursorResult(
                sequence: Self.clampExpired(cursorSequence, firstSequence),
                expired: Self.isExpired(cursorSequence, firstSequence))
        }

        if isHistoryQuery {
            return CursorResult(
                sequence: firstSequence == 0 ? 0 : max(0, firstSequence - 1),
                expired: false)
        }

        let key = Self.normalizeClientId(clientId).lowercased()
        if var entry = clientPositions[key] {
            // touch: 読み出しでも lastSeen を更新しないと、subscribe しっぱなしで
            // advance だけ呼ばれないパス (= 取得 0 件で位置が進まない) が TTL で落ちる。
            entry.lastSeen = now
            clientPositions[key] = entry
            return CursorResult(
                sequence: Self.clampExpired(entry.sequence, firstSequence),
                expired: Self.isExpired(entry.sequence, firstSequence))
        }

        clientPositions[key] = Entry(sequence: latestSequence, lastSeen: now)
        return CursorResult(sequence: latestSequence, expired: false)
    }

    public func advance(clientId: String, sequence: Int64, now: Date) {
        clientPositions[Self.normalizeClientId(clientId).lowercased()] =
            Entry(sequence: sequence, lastSeen: now)
    }

    /// 最終アクセスから ttl 経過した client entry を削除する。戻り値は削除件数。
    @discardableResult
    public func pruneStale(now: Date, ttl: TimeInterval? = nil) -> Int {
        let threshold = now.addingTimeInterval(-(ttl ?? Self.defaultStaleClientTtl))
        let stale = clientPositions.filter { $0.value.lastSeen < threshold }.map(\.key)
        for key in stale {
            clientPositions.removeValue(forKey: key)
        }
        return stale.count
    }

    public var trackedClientCount: Int { clientPositions.count }

    public static func encode(_ sequence: Int64) -> String {
        Base64Url.encode(Data("seq:\(sequence)".utf8))
    }

    private static func clampExpired(_ sequence: Int64, _ firstSequence: Int64) -> Int64 {
        isExpired(sequence, firstSequence) ? max(0, firstSequence - 1) : sequence
    }

    private static func isExpired(_ sequence: Int64, _ firstSequence: Int64) -> Bool {
        firstSequence > 0 && sequence < firstSequence - 1
    }

    private static func normalizeClientId(_ clientId: String) -> String {
        let trimmed = clientId.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "anonymous" : trimmed
    }

    static func decode(_ cursor: String) -> Int64? {
        guard !cursor.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let data = Base64Url.decode(cursor),
              let text = String(data: data, encoding: .utf8),
              text.lowercased().hasPrefix("seq:") else {
            return nil
        }
        return Int64(text.dropFirst(4))
    }
}
